PORTAL/jobs/_utils.py

```diff
@@ -3727,14 +3724,15 @@ class SSHCommands:
 
     @classmethod
     def from_config(cls, cfg, **kwargs):
-        return cls(cfg.user, cfg.host, cfg.port, **kwargs)
+        return cls(cfg.user, cfg.host, cfg.port, cfg.agent, **kwargs)
 
-    def __init__(self, user, host, port, *, ssh=None, scp=None):
+    def __init__(self, user, host, port, agent, *, ssh=None, scp=None):
         self.user = check_shell_str(user)
         self.host = check_shell_str(host)
         self.port = int(port)
         if self.port < 1:
             raise ValueError(f'invalid port {self.port}')
+        self.agent = agent
```

Comment:
Think of `SSHClient` (and `SSHCommands`, etc.) as a concrete wrapper around a configured endpoint.  You don't need auth info until you go to connect.  That's why the agent info is optionally passed in to the various methods instead of being an attribute.

I'm not opposed to binding the agent info to `SSHClient` (etc.), but consider the current usage:

* `SSHClient` (etc.) encapsulates its info fairly completely
* the agent info doesn't fit that idea very well; it is used independently of any specific connection info, in various capacities
 * in `Job._get_ssh_agent()` we calculate the agent info to use, based on the config (which shouldn't change) and the env (which could)
 * `Job._get_ssh_agent()` is called fresh each time we need to use agent info
 * since we won't necessarily be using the configured agent, the `SSHClient` (etc.) methods must take the agent as an argument

The question is if `self.worker.ssh.agent` should be valid.  It would probably mean "the agent to use when connecting".  That does make some sense.  If we go that route, we would likely do the following:

1. respect the "agent" parameter of the several `SSHClient` (etc.) methods as an override (use `self.agent` as the default)
2. instead of the configured info, make `agent` a property that computes the agent each time using the implementation from `Job._get_ssh_agent()`

We should also consider storing the original `SSHConnectionConfig` on the `SSHClient` (etc.) instance.  Then we could look up `self.worker.ssh.cfg.agent`.

If we do not make the agent info an integral part of `SSHClient` (etc.) then we should consider separating the agent config from the worker SSH config.  However, having written that down, that idea doesn't make a lot of sense.

Reply:
> We should also consider storing the original `SSHConnectionConfig` on the `SSHClient` (etc.) instance. Then we could look up `self.worker.ssh.cfg.agent`.

That might be the simplest thing to do (aside from leaving `SSHClient`, etc. alone).  We could circle back later if it were important.

Reply:
Yeah, just removed this for now because it is possible to get this from the config object as it did before.

Reply:
Cool.